consensus/cometbft/service/finalize_block.go

```diff
@@ -48,52 +54,91 @@ func (s *Service) finalizeBlockInternal(
 		return nil, err
 	}
 
-	// finalizeBlockState should be set on InitChain or ProcessProposal. If it
-	// is nil, it means we are replaying this block and we need to set the state
-	// here given that during block replay ProcessProposal is not executed by
-	// CometBFT.
-	if s.finalizeBlockState == nil {
-		s.finalizeBlockState = s.resetState(ctx)
-	} else {
-		// Preserve the CosmosSDK context while using the correct base ctx.
-		s.finalizeBlockState.SetContext(s.finalizeBlockState.Context().WithContext(ctx))
+	// Check whether currently block hash is already available. If so
+	// we may speed up block finalization.
+	hash := string(req.Hash)
+	switch cached, err := s.cachedStates.GetCached(hash); {
+	case err == nil:
+		// Block with height equal to initial height is special and we can't rely on its cache.
+		// This is because Genesis state is cached but not committed (and purged from s.cachedStates)
+		// We handle the case below.
+		if req.Height > s.initialHeight {
+			if err = s.cachedStates.MarkAsFinal(hash); err != nil {
+				return nil, fmt.Errorf("failed marking state as final, hash %s, height %d: %w", hash, req.Height, err)
+			}
+
+			finalState := cached.State
+			var (
+				signedBlk *ctypes.SignedBeaconBlock
+				sidecars  datypes.BlobSidecars
+			)
+			signedBlk, sidecars, err = s.Blockchain.ParseBeaconBlock(req)
+			if err != nil {
+				return nil, fmt.Errorf("finalize block: failed parsing block: %w", err)
+			}
+			blk := signedBlk.GetBeaconBlock()
+
+			if err = s.Blockchain.FinalizeSidecars(
+				finalState.Context(),
+				req.SyncingToHeight,
+				blk,
+				sidecars,
+			); err != nil {
+				return nil, fmt.Errorf("failed finalizing sidecars: %w", err)
+			}
+			if err = s.Blockchain.PostFinalizeBlockOps(
+				finalState.Context(),
+				blk,
+			); err != nil {
+				return nil, fmt.Errorf("finalize block: failed post finalize block ops: %w", err)
+			}
+			return formatFinalizeBlockResponse(cached.ValUpdates, len(req.Txs), s.cmtConsensusParams)
```

Comment:
This is the fast track to finalize blocks already verified. We just store blobs/block, send FCU and pull deposits to be included in next block